Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp

```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
 {
     return webkitFeatureListCreate(WebPreferences::internalDebugFeatures());
 }
+
+/**
+ * webkit_settings_apply_from_key_file:
+ * @settings: a #WebKitSettings
+ * @key_file: a #GKeyFile
+ * @group_name: Name of the group to read from @key_file
+ * @error: return location for error or %NULL to ignore
+ *
+ * Reads the contents of the given @group_name from the given @key_file and apply the value of
+ * each key/value to the corresponding property on the @settings.
```

Comment:
I'm not sure that's exactly what this does, or I'm not understanding it. We iterate the WebKitSetting properties, not the key file, failing if the key file doesn't have a property, which requires to load all properties?

Reply:
We should also document how can this fail and what happens when it fails. It seems this is not atomic, so if we fail to set a property, the previous ones are set anyway, but we stop so the next one won't even if they are valid.

Reply:
That algorithm was Adrian's, I didn't change it.

Maybe it'd be simpler to iterate on the key/values instead?

Reply:
@carlosgcampos If the key file does not have an entry for a property, it's skipped (see comment in line 4171). Properties are iterated to know which key names to check. It could also be done the other way around: iterating over the existing keys and looking up the corresponding property by name.

Regarding atomicity, I suppose if we wanted to make sure all applied settings are valid we are probably going to need two passes, one to validate and type-check the values in the `GKeyFile`, storing them in an arrays of names and `GValue`s, and then applying them all at once using [g_object_setv()](https://docs.gtk.org/gobject/method.Object.setv.html)... Maybe stuff for a follo-up?

Reply:
> @carlosgcampos If the key file does not have an entry for a property, it's skipped (see comment in line 4171). Properties are iterated to know which key names to check. It could also be done the other way around: iterating over the existing keys and looking up the corresponding property by name.

I've seen the comment, but I don't understand how it works. It seems we iterate the properties and fail we fail to find it in the key file. Am I understanding the code wrong?

> Regarding atomicity, I suppose if we wanted to make sure all applied settings are valid we are probably going to need two passes, one to validate and type-check the values in the `GKeyFile`, storing them in an arrays of names and `GValue`s, and then applying them all at once using [g_object_setv()](https://docs.gtk.org/gobject/method.Object.setv.html)... Maybe stuff for a follow-up?

We are not in a hurry to land this, I prefer to land a version that does what we want with tests checking those cases.

Reply:
I wrote this a good while ago, but trying to think like myself in the past, probably the reasoning was “I will need the `GParamSpec` for the property anyway, so if I iterate over the properties I already get that _and_ the name at once”.

Feel free to change it, but repeatedly calling `g_object_find_property()` will do a linear search for each property.

Reply:
> > @carlosgcampos If the key file does not have an entry for a property, it's skipped (see comment in line 4171). Properties are iterated to know which key names to check. It could also be done the other way around: iterating over the existing keys and looking up the corresponding property by name.
> 
> I've seen the comment, but I don't understand how it works. It seems we iterate the properties and fail we fail to find it in the key file. Am I understanding the code wrong?

The algorithm is basically:

```
  for each *property* of the object class:
     if *keyfile* errored while checking if there's a key named after *property*:
        return with error

     if *keyfile* did _not_ contain a key named after *property*:
        skip to the next property

    convert the value for the key to the property type
    if the conversion suceeded:
       set property to the convered value
    else:
       return with error
```

I hope this helps.

> > Regarding atomicity, I suppose if we wanted to make sure all applied settings are valid we are probably going to need two passes, one to validate and type-check the values in the `GKeyFile`, storing them in an arrays of names and `GValue`s, and then applying them all at once using [g_object_setv()](https://docs.gtk.org/gobject/method.Object.setv.html)... Maybe stuff for a follow-up?
> 
> We are not in a hurry to land this, I prefer to land a version that does what we want with tests checking those cases.

Sure.

Reply:
Ah! it makes sense now, thanks for the clarification. It's weird a function returning FALSE and not setting error, but it's a query function.